constraint-solver/src/grouped_expression.rs

```diff
@@ -231,6 +231,28 @@ impl<T: RuntimeConstant, V: Ord + Clone + Eq> GroupedExpression<T, V> {
         }
     }
 
+    /// Splits this expression into head and tail, i.e., `self = head + tail`
+    /// head is the first summand
+    pub fn try_split_head_tail(&self) -> Option<(Self, Self)> {
+        if self.linear_components().len() + self.quadratic_components().len() < 2 {
+            return None;
+        }
+
+        let head = self.clone().into_summands().next()?;
+        let head = match head {
+            GroupedExpressionComponent::Quadratic(l, r) => l * r,
+            GroupedExpressionComponent::Linear(var, coeff) => {
+                GroupedExpression::from_runtime_constant(coeff)
+                    * GroupedExpression::from_unknown_variable(var)
+            }
+            GroupedExpressionComponent::Constant(_) => {
+                return None;
+            }
+        };
+        let tail = self.clone() - head.clone();
```

Comment:
If possible, I think we sholud try to avoid this subtraction here. Subtraction is relatively expensive, especially if quadratic terms are involved (because they need to be matched against each other).

Since we are inside GroupedExpression here, it is OK to make assumptions about the internal representation, i.e. we can do things like
```rust
if !self.quadratic.is_empty() {
  let mut quadratic = self.quadratic.into_iter();
  let (hl, hr) = self.quadratic.next().unwrap();
  self.quadratic = quadratic.collect();
  Some((hl * hr), self))
} ...
```